### PR TITLE
Resolve race condition preventing last summary from generating

### PR DESCRIPTION
**Pull request by arinwt**
This is part of #2082.

There is a race condition preventing the "last summary" before closing from getting a chance to generate/submit.

1. As soon as summarizer client starts, it will try to generate a summary with reason "maxOps"
1. Then it will get the stop request from the parent client disconnecting, which will await trySummarize
1. The broadcast deferred returned from trySummarize will be an immediately resolved promise
1. The stop flow will then call the container closeFn which will disconnect it
1. Generating summary will abort, because it checks if still connected between each await step

This fixes in two parts:
- The race condition is resolved by instead awaiting the promise of `this.summarizing`, which is a deferred whose existence serves to prevent re-entrancy as well as resolve when the active summarizing step completes.
- The deferred now resolves on full completion rather than on seeing its summarize op be broadcast. This is important in case the "last summary" needs to be a "safe" summary. It also may help telemetry by showing a successful summary if it actually succeeds, rather than showing as a canceled performance event.

### CODE REVIEW

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 1]
```diff
@@ -387,36 +389,33 @@ export class RunningSummarizer implements IDisposable {
...
-            this.summarizing = false;
-            broadcastDeferred.resolve();
+            this.summarizing?.resolve();
+            this.summarizing = undefined;
             if (this.tryWhileSummarizing) {
                 this.tryWhileSummarizing = false;
                 this.heuristics.run();
             }
         }).catch((error) => {
             this.logger.sendErrorEvent({ eventName: "UnexpectedSummarizeError" }, error);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:417]: I think we want to reject the promise here, no?

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 2]
```diff
@@ -387,36 +389,33 @@ export class RunningSummarizer implements IDisposable {
...
     }
 
-    private trySummarize(reason: string): { broadcastP: Promise<void> } {
-        if (this.summarizing === true) {
+    private trySummarize(reason: string): void {
+        if (this.summarizing !== undefined) {
             // We can't summarize if we are already
             this.tryWhileSummarizing = true;
-            return { broadcastP: Promise.resolve() };
+            return;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:396]: It feels to me that you can make this.summarizing be scoped only to this function by returning this.summarizing.promise here, and thus making code in waitStop() await this.trySummarize() call.

**Review by vladsud — APPROVED** [review]

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 1 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:417]: Why reject?

Comment (vladsud) [reply packages/runtime/container-runtime/src/summarizer.ts:417]: Oh, it's fire-and-forget? Sorry, missed that, then yes, we do not care.

**Pull request merged**